inference-engine/src/auto_plugin/auto_exec_network.hpp

```diff
@@ -12,36 +12,153 @@
 #include <vector>
 #include <string>
 
+#include <ie_parallel.hpp>
 #include <cpp_interfaces/impl/ie_executable_network_thread_safe_default.hpp>
 #include <threading/ie_itask_executor.hpp>
 
+#if (IE_THREAD == IE_THREAD_TBB || IE_THREAD == IE_THREAD_TBB_AUTO)
+# include <tbb/concurrent_queue.h>
+#endif
+
 namespace AutoPlugin {
 
+class AutoInferencePlugin;
+
 using DeviceName = std::string;
+using ConfigType = std::map<std::string, std::string>;
 using NetworkFuture = std::future<InferenceEngine::SoExecutableNetworkInternal>;
 
-class AutoExecutableNetwork : public InferenceEngine::IExecutableNetworkInternal {
+template<typename T>
+using DeviceMap = std::unordered_map<DeviceName, T>;
+
+#if ((IE_THREAD == IE_THREAD_TBB) || (IE_THREAD == IE_THREAD_TBB_AUTO))
+template <typename T>
+using ThreadSafeQueue = tbb::concurrent_queue<T>;
+template <typename T>
+using ThreadSafeBoundedQueue = tbb::concurrent_bounded_queue<T>;
+#else
+template <typename T>
+class ThreadSafeQueue {
+public:
+    void push(T value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        _queue.push(std::move(value));
+    }
+    bool try_pop(T& value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        if (!_queue.empty()) {
+            value = std::move(_queue.front());
+            _queue.pop();
+            return true;
+        } else {
+            return false;
+        }
+    }
+protected:
+    std::queue<T>   _queue;
+    std::mutex      _mutex;
+};
+template <typename T>
+class ThreadSafeBoundedQueue {
+public:
+    ThreadSafeBoundedQueue() = default;
+    bool try_push(T value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        if (_capacity > _queue.size()) {
+            _queue.push(std::move(value));
+            return true;
+        }
+        return false;
+    }
+
+    void push(T value) {
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            while (_capacity <= _queue.size()) {
+                _notFull.wait(lock);
+            }
+            _queue.push(std::move(value));
+        }
+        _notEmpty.notify_one();
+    }
+
+    void pop(T& value) {
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            while (!_capacity || _queue.empty()) {
+                _notEmpty.wait(lock);
+            }
+            value = std::move(_queue.front());
+            _queue.pop();
+        }
+        _notFull.notify_one();
+    }
+
+    bool try_pop(T& value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        if (_capacity && !_queue.empty()) {
```

Comment:
Why `_capacity` is involved here?

Reply:
I guess this is to avoid the situation of forgetting set_capacity(), I just copy this part from [L77](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/multi_device/multi_device_exec_network.hpp#L77) , but I agree with you, I will remove those redundent code, thanks!